### PR TITLE
NTPSec

### PR DESCRIPTION
**Pull request by georgyo**
###### Motivation for this change

NTPSec is fork of ntpd, and there for a  drop in replacement of it. It includes numerous security fixes and improvements, as well as implementations of secure time protocols such as NTS.

Instead of creating a new ntpsec service configuration, I decided it would be better to just reuse the existing ntp configuration, and add a few more customizations to that that one.

This is my first time trying to upstream a nix package, please feel free to point out anything I can do better.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions (archlinux)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by jonringer — COMMENTED** [review]
To conform to CONTRIBUTING.md, you'll probably want to chagne your commit messages to similar to:
```
maintainers: add georgyo
ntpsec: init at 1.1.4
nixos/networking: add additional ntp options
```
I'm not 100% how to label nixos changes, but I think that would be more accurate and inline with CONTRIBUTING.md

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [inline pkgs/tools/networking/ntpsec/default.nix]: libcap should be moved to the optional line
```suggestion
  buildInputs = [ libbsd bison openssl_1_1 python ]
```

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [inline pkgs/tools/networking/ntpsec/default.nix]:
```suggestion
    ++ lib.optionals stdenv.isLinux [ libcap pps-tools ];
```

**Comment by georgyo** [comment]
Thanks, commit messages are now fixed up.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [inline pkgs/tools/networking/ntpsec/default.nix]: You can list many
```suggestion
    license = with licenses; [ bsd2 bsd3 mit ];
```
for completeness, you can add the ntp license, if you want.

**Review by jonringer — COMMENTED** [review]
i think this libcaps logic is more in line with what you're trying to express.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (georgyo) [reply pkgs/tools/networking/ntpsec/default.nix]: This was copying the how it was done in ntp, the check is done in all-packages.nix. I am not sure why it was done this way, but indeed if I try doing it this way it does not get linked. I cannot explain why it does not get linked if configured this way.

Commit 458c8381e03b is what changes where this gets checked for ntp. However it's commit message is lacking as it implies only a version bump.

File: nixos/modules/services/networking/ntpd.nix
Comment (markuskowa) [inline nixos/modules/services/networking/ntpd.nix]: I would simply call this option just `package` as many other modules do (ntp is then also included in the toplevel attribute: `config.services.ntp.package`).

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [reply pkgs/tools/networking/ntpsec/default.nix]: hmm, I think it's a matter of personal taste, but I personally like to see all logic in the file that pertains to how to construct that package. 

If you were going to offer [many variations of a package](https://github.com/NixOS/nixpkgs/blob/6ea5cf34b211b9ddcdb8a30cead68ee461b65c13/pkgs/top-level/python-packages.nix#L5408), this would make sense to me. But in this case, it seems all of it could be put into the nix file itself.

I don't feel super strong about this, but I'm sure someone else might :)

Comment (georgyo) [reply pkgs/tools/networking/ntpsec/default.nix]: I actually like your way better, but I couldn't get it to work... Though I finally figured out that it was `optionals` and not just `optional` when putting an array after the check... Updated.

File: nixos/modules/services/networking/ntpd.nix
Comment (georgyo) [reply nixos/modules/services/networking/ntpd.nix]: good idea, fixed.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (JohnAZoidberg) [reply pkgs/tools/networking/ntpsec/default.nix]: `optionals` takes a list and returns either that list or the empty list.
`optional` takes an element and returns either that element wrapped in a list or the empty list.

Yes, I also think it's better to have all the logic in the package file, since it's not really a personal choice here but a platform choice that changes how to build the package.

File: nixos/modules/services/networking/ntpd.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/networking/ntpd.nix]: Nitpick: You don't need the double-single-quotes for a single line.
```suggestion
        example = literalExample "pkgs.ntpsec";
```

File: nixos/modules/services/networking/ntpd.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/networking/ntpd.nix]: In line 9 it's aliased to `cfg` like (pretty much) all modules do it. Then it's immediately obvious that the config comes from the current package and not another one.

```suggestion
    environment.systemPackages = [ cfg.package ];
```

File: nixos/modules/services/networking/ntpd.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/networking/ntpd.nix]:
```suggestion
          ExecStart = "@${cfg.package}/bin/ntpd ntpd -g ${ntpFlags}";
```

File: pkgs/tools/networking/ntpsec/default.nix
Comment (JohnAZoidberg) [inline pkgs/tools/networking/ntpsec/default.nix]: If the platform is Linux only, do you have to do the checks and optional expressions for Linux?

File: pkgs/tools/networking/ntpsec/default.nix
Comment (JohnAZoidberg) [inline pkgs/tools/networking/ntpsec/default.nix]:
```suggestion
    homepage = "https://www.ntpsec.org/";
```

We're probably going to deprecate unquoted strings soon, once RFC 45 is approved.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [reply pkgs/tools/networking/ntpsec/default.nix]: Yep, not sure where to find the documentation for lib.optional, but luckily it's [also just in a plain nix file](https://github.com/NixOS/nixpkgs/blob/b39f2efa182ea6e3d1be2a0e9e8a81771fdf4181/lib/lists.nix#L241).

File: pkgs/tools/networking/ntpsec/default.nix
Comment (jonringer) [reply pkgs/tools/networking/ntpsec/default.nix]: The package itself can be built for most [unix platforms](https://docs.ntpsec.org/latest/build.html#unix). I think the author is just trying to say it's only verified on linux for now.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (JohnAZoidberg) [reply pkgs/tools/networking/ntpsec/default.nix]: In the [nixpkgs manual](https://nixos.org/nixpkgs/manual/#function-library-lib.lists.optional)

File: pkgs/tools/networking/ntpsec/default.nix
Comment (georgyo) [reply pkgs/tools/networking/ntpsec/default.nix]: This is another thing copied from the [ntp package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/networking/ntp/default.nix). Which seemed like a good starting point for NTPsec.

As @jonringer said,  NTPsec is tested on a great many platforms, including macos and freebsd as well as both big endian and little endian systems. The full list of upstream tests can be found here: https://gitlab.com/NTPsec/ntpsec/blob/master/.gitlab-ci.yml

So while it should work fine on other systems,  I only have the ability to test on linux.

File: nixos/modules/services/networking/ntpd.nix
Comment (georgyo) [reply nixos/modules/services/networking/ntpd.nix]: Good idea!

File: nixos/modules/misc/ids.nix
Comment (ajs124) [inline nixos/modules/misc/ids.nix]: Is this necessary? Can't the module use DynamicUsers?

Comment (georgyo) [reply nixos/modules/misc/ids.nix]: I don't think we can use DynamicUsers here. This is not a new module, so it must maintain interoperability with configurations out there. While NTP might seem like it would not need to write anything to disk, it does. So changing the uid now would make NTP unable to read those files.

Also, this would be the first module in nixpkgs to use DynamicUsers.  I don't want to break the mold on my first MR.

To the question if this is necessary, as this is for the gid and not the uid anyway. I believe it to be good hygiene. NTP does write to disk, and under the old configuration those files would be owned by ntp:nogroup. This both does not look good, and grants other unprivileged "nogroup" services extra surface area on files ntp writes.

I have verified that changing the gid for NTP will not break anything as it also owns all the files it is using.

Comment (JohnAZoidberg) [reply nixos/modules/misc/ids.nix]: It seems that a lot of modules are using `DynamicUser` (notice the missing s at the end) already:

```
$ git grep 'DynamicUser = true' | wc -l
45
```

Do those files that NTP writes need to persist across restarts or are they temporary files? Because if they are then it wouldn't hurt losing them.

Maybe you could also use `StateDirectory`, ` RuntimeDirectory`or `tmpfiles.d` or the likes for the temp files instead of `/var/lib/ntp`. See https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html and https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=

Comment (georgyo) [reply nixos/modules/misc/ids.nix]: You are right about current usage... I am really getting burned by the letter `s` this MR. I did the grep for `DynamicUsers`. However my other points stand.

These files are not temporary, for ntpd this is namely the drift file is used.
>The drift file is used to store the frequency offset between the system clock running at its nominal frequency and the frequency required to remain in synchronization with UTC. If present, the value contained in the drift file is read at system start and used to correct the clock source. Use of the drift file reduces the time required to achieve a stable and accurate time. The value is calculated, and the drift file replaced, once per hour by ntpd. The drift file is replaced, rather than just updated, and for this reason the drift file must be in a directory for which ntpd has write permissions. 

For ntpsec there are can be several stats files that are meant to be read by other programs.

But beyond that, this is a large change to an existing core NTP module. If we want to go down the route to moving to DynamicUser then it should be in it's own CR, not complicate this one.

Comment (georgyo) [reply nixos/modules/misc/ids.nix]: I have removed the group bits here, which trims down the changes to the module. I am not opposed to switching to `DynamicUser` and making the service unit file much more explicit about what it can and cannot do. However the main goal of this MR is just adding NTPsec.

After it get's merged I'll create a MR for reworking this module a bit. But unlike this change of just adding a package, that change will require significantly more testing.

**Comment by georgyo** [comment]
Anything blocking this from getting merged now?

File: nixos/modules/services/networking/ntpd.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/networking/ntpd.nix]: Is this a good idea?

File: nixos/modules/services/networking/ntpd.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/networking/ntpd.nix]:
```suggestion
        default = pkgs.ntp;
        defaultText = "pkgs.ntp";
```

File: nixos/modules/services/networking/ntpd.nix
Comment (georgyo) [reply nixos/modules/services/networking/ntpd.nix]: Yes, if you care about a have a highly accurate system clock. A system under high load will have scheduling delays, which leads to increase jitter. I have never worked in a place that hasn't had -N.

Though, you are right that this can just be added to extraFlags, further reducing this diff.

Comment (JohnAZoidberg) [reply nixos/modules/services/networking/ntpd.nix]: You don't need to minimize the diff. I think it's nice if NixOS has great, pleasantly surprising default configs. If you think 90% of people would want to use it with `-N` then please add it :)

I just wanted to make sure it's a good idea since the manual says it will result in ntpd being scheduled with a higher priority than the other processes - so that might affect system performance. (Which is probably of little disturbance with ntpd - but I wouldn't know).

If you think it should be the default but you should be able to turn it off  easily, add a config option for it and have it default to `true`.

**Comment by georgyo** [comment]
Version bumped to 1.1.6.

**Comment by georgyo** [comment]
Version bumped to 1.1.7.

File: pkgs/tools/networking/ntpsec/default.nix
Comment (ryantm) [inline pkgs/tools/networking/ntpsec/default.nix]:
```suggestion
  pname = "ntpsec";
```

**Review by ryantm — CHANGES_REQUESTED** [review]

**Comment by ryantm** [comment]
@georgyo Sorry this is taking so long to get in. It looks good to me except for that one pname thing. Could you fix that and we should be good to go?

**Comment by ryantm** [comment]
@GrahamcOfBorg  build ntpsec

**Comment by ryantm** [comment]
Looks like it is failing the OfBorg build check on x86-64 too.

**Comment by georgyo** [comment]
Interesting. I think this is failing to build with gcc9 which is now on master. I'll take a deeper look this weekend.

**Pull request closed without merging**
